src/server/gc_worker/compaction_filter.rs

```diff
@@ -398,6 +401,34 @@ impl WriteCompactionFilter {
         }
     }
 
+    fn do_filter_new_key(
+        &mut self,
+        mvcc_key: &[u8],
+        write: WriteRef<'_>,
+    ) -> Result<CompactionFilterDecision, String> {
+        self.handle_bottommost_delete();
+        self.switch_key_metrics();
+        self.mvcc_key_prefix.clear();
+        match write.write_type {
+            WriteType::Rollback | WriteType::Lock => {
+                self.mvcc_rollback_and_locks += 1;
+                self.filtered += 1;
+                return Ok(CompactionFilterDecision::Remove);
```

Comment:
So in this case the current key (`self.mvcc_key_prefix`) is actually not switched (but set to empty), and when we found the next version of the same user key, this function will be called again, and `handle_bottommost_delete` and `switch_key_metrics` are also called again?

I'm afraid this doesn't seem to make the code clearer and easier to read...  The `remove_older` flag still looks necessary I think